### DISCUSSION TITLE
How to use macros with mhchem

### DISCUSSION THREAD

**Discussion by jiaopengzi** [KaTeX/KaTeX#3915] [Q&A] [answered]
```html
<!-- use macros with mhchem -->
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
    <title>Marked with mhchem</title>
    <link
      rel="stylesheet"
      href="https://cdn.jsdelivr.net/npm/katex@0.16.9/dist/katex.min.css"
    />
    <script src="https://cdn.jsdelivr.net/npm/katex@0.16.9/dist/katex.min.js"></script>
    <script src="https://cdn.jsdelivr.net/npm/katex@0.16.9/dist/contrib/mhchem.min.js"></script>
  </head>
  <body>
    <div id="content"></div>

    <script>
      const options = {
        throwOnError: false,
        displayMode: true,
        macros: {
          '\\ce': '\\mhchem{#1}',
        },
      }
      const str =
        '\\ce{N2 + 3H2 <=>T[High temperature Pressurized][Catalyst] 2NH3}'
      document.getElementById('content').innerHTML = katex.renderToString(
        str,
        options
      )
    </script>
  </body>
</html>

```
![image](https://github.com/KaTeX/KaTeX/assets/52224629/882e5934-1213-4f44-a403-a98717d66a55)

```html
<!-- only use katex -->
<!DOCTYPE html>
<html>
  <head>
    <link
      rel="stylesheet"
      href="https://cdn.jsdelivr.net/npm/katex@0.16.9/dist/katex.min.css"
    />
    <script
      defer
      src="https://cdn.jsdelivr.net/npm/katex@0.16.9/dist/katex.min.js"
    ></script>
    <script
      defer
      src="https://cdn.jsdelivr.net/npm/katex@0.16.9/dist/contrib/mhchem.min.js"
    ></script>
    <script
      defer
      src="https://cdn.jsdelivr.net/npm/katex@0.16.9/dist/contrib/auto-render.min.js"
      onload="renderMathInElement(document.body);"
    ></script>
  </head>
  <body>
    <span>
      $$\ce{N2 + 3H2 <=>T[High temperature Pressurized][Catalyst] 2NH3}$$
    </span>
  </body>
</html>

```
![image](https://github.com/KaTeX/KaTeX/assets/52224629/5fc9908e-37f2-483a-9b92-e85b6bd1c0ae)

- only use katex is ok
- use macros with mhchem is not ok

Hello,

How can I use mhchem macros in the options to parse chemical formulas properly? I saw in the official documentation that it can be directly included, but I am having trouble getting it to work with Vue 3. Could you please provide me some assistance in making "use macros with mhchem" function correctly? I greatly appreciate your help.

**Comment by edemaine** [comment]
Your macro `'\\ce': '\\mhchem{#1}'` is overwriting the default notion of `\ce` to instead call `\mhchem` which isn't defined. You should be able to define macros in the way you have shown; you just don't want to modify macro `\ce` because it works as is.

**Reply by jiaopengzi** [reply]
Very good, thank you for your answer. I may not have a deep understanding of the underlying mechanism of KaTeX macros.

How can I modify the HTML using `use macros with mhchem` to achieve the effect of `only use katex` in the HTML?

Please kindly help me modify the example of the HTML using `use macros with mhchem` to achieve the desired effect.

Thank you very much.

**Reply by jiaopengzi** [reply]
```html
<!-- use marked  -->
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
    <title>Marked in the browser</title>
    <link
      rel="stylesheet"
      href="https://cdn.jsdelivr.net/npm/katex@0.16.9/dist/katex.min.css"
    />
    <script src="https://cdn.jsdelivr.net/npm/katex@0.16.9/dist/katex.min.js"></script>
    <script src="https://cdn.jsdelivr.net/npm/katex@0.16.9/dist/contrib/mhchem.min.js"></script>
    <script src="https://cdn.jsdelivr.net/npm/marked-katex-extension/lib/index.umd.js"></script>
    <script src="https://cdn.jsdelivr.net/npm/marked/marked.min.js"></script>
  </head>
  <body>
    <div id="content"></div>

    <script>
      const options = {
        throwOnError: false,
        displayMode: true,
        macros: {
          '\\ce': '\\mhchem{#1}',
        },
      }

      marked.use(markedKatex(options))
      document.getElementById('content').innerHTML = marked.parse(
        '$$\\ce{N2 + 3H2 <=>T[High temperature Pressurized][Catalyst] 2NH3}$$'
      )
    </script>
  </body>
</html>

```



Previously, I simplified the problem to the `katex` level. In fact, my actual requirement is that `marked` needs to be compatible with parsing chemical formulas during its analysis.

I have noticed the [mhchem](https://github.com/KaTeX/KaTeX/tree/main/contrib/mhchem) documentation, but there is no clearer guide on how to integrate it with marked's options for parsing chemical formulas.

I feel my logic is correct, but I am unsure about the implementation details.

**Reply by jiaopengzi** [reply]
@edemaine 
Hello, I have added some content, could you please give me some guidance?

**Reply by edemaine** [reply]
Looking at the code for marked-katex-extension, I think you need to make sure it's importing the same `katex` as you are, so that mhchem installs its macros correctly. I'm not sure how to do that with your CDN setup...

**Reply by jiaopengzi** [reply, marked as answer]
thanks all. it is working. see [here](https://github.com/UziTech/marked-katex-extension/issues/315)
